server/services/sequence_service_test.go

```diff
@@ -0,0 +1,223 @@
+	s := NewSequenceServer()
+
+	seq, err := s.CreateSequence(context.Background(), &pb.CreateSequenceRequest{})
+	if err != nil {
+		t.Errorf("CreateSequence(out_of_range): unexpected err %+v", err)
+	}
+
+	timeout := 5 * time.Second
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	e, err := s.AttemptSequence(ctx, &pb.AttemptSequenceRequest{Name: seq.GetName()})
+	if err != nil {
+		t.Errorf("AttemptSequence(out_of_range): unexpected err %+v", err)
+	}
+
+	if e == nil {
+		t.Errorf("AttemptSequence(out_of_range): unexpected nil Empty response")
+	}
+
+	_, err = s.AttemptSequence(ctx, &pb.AttemptSequenceRequest{Name: seq.GetName()})
+	if c := status.Code(err); c != codes.OutOfRange {
+		t.Errorf("%s: status was %v wanted %v", t.Name(), c, codes.OutOfRange)
+	}
+
+	_, err = s.AttemptSequence(ctx, &pb.AttemptSequenceRequest{Name: seq.GetName()})
+	if c := status.Code(err); c != codes.OutOfRange {
+		t.Errorf("%s: status was %v wanted %v", t.Name(), c, codes.OutOfRange)
+	}
+
+	r := report(seq.GetName())
+	report, err := s.GetSequenceReport(context.Background(), &pb.GetSequenceReportRequest{Name: r})
+	if err != nil {
+		t.Errorf("GetSequenceReport(out_of_range): unexpected err %+v", err)
+	}
+
+	attempts := report.GetAttempts()
+	if len(attempts) != 3 {
+		t.Errorf("%s: expected number of attempts to be 3 but was %d", t.Name(), len(attempts))
+	}
+
+	a := attempts[0]
+	ad := a.GetAttemptDeadline().AsTime()
+	d, _ := ctx.Deadline()
+
+	if !ad.Equal(d) {
+		t.Errorf("%s: server deadline = %v client deadline = %v", t.Name(), ad, d)
+	}
+}
+
+func TestGetSequenceReportNotFound(t *testing.T) {
+	s := NewSequenceServer()
+	_, err := s.GetSequenceReport(context.Background(), &pb.GetSequenceReportRequest{Name: "foo/bar/baz"})
+	if c := status.Code(err); c != codes.NotFound {
+		t.Errorf("%s: expected error to be %s but was %s", t.Name(), codes.NotFound.String(), c.String())
+	}
+}
+
+func TestGetSequenceReportMissingName(t *testing.T) {
+	s := NewSequenceServer()
+	_, err := s.GetSequenceReport(context.Background(), &pb.GetSequenceReportRequest{Name: ""})
+	if c := status.Code(err); c != codes.InvalidArgument {
+		t.Errorf("%s: expected error to be %s but was %s", t.Name(), codes.InvalidArgument.String(), c.String())
+	}
+}
+
+func TestAttemptSequenceNotFound(t *testing.T) {
+	s := NewSequenceServer()
+	_, err := s.AttemptSequence(context.Background(), &pb.AttemptSequenceRequest{Name: "foo/bar/baz"})
+	if c := status.Code(err); c != codes.NotFound {
+		t.Errorf("%s: expected error to be %s but was %s", t.Name(), codes.NotFound.String(), c.String())
+	}
+}
+
+func TestAttemptSequenceMissingName(t *testing.T) {
+	s := NewSequenceServer()
+	_, err := s.AttemptSequence(context.Background(), &pb.AttemptSequenceRequest{Name: ""})
+	if c := status.Code(err); c != codes.InvalidArgument {
+		t.Errorf("%s: expected error to be %s but was %s", t.Name(), codes.InvalidArgument.String(), c.String())
```

Comment:
nit: Do you need to call `String()` explicitly in all these places? Don't the *f functions do that automatically if the arg satisfies the string interface?

Reply:
Ah you are right, I removed them. Done.